dataplane/README.md

```diff
@@ -0,0 +1,155 @@
+
+Note: If no actions are added to a port: all packets are dropped!
+
+Example ports types:
+
+1. Kernel: uses linux AF_PACKET
+2. TAP: uses ioctl and a file descriptor to create/read to TAP interface
+
+Implementation: [fwdport](forwarding/fwdport/ports)
+
+#### Tables
+
+Tables match against packet headers and perform actions. Each table has a default action list, and list of entries. An entry defines a set of field and values to match against
+and a list of the corresponding actions.
+
+Note: After a packet is looked up in a table, the new actions are processed before any pending actions.
+
+Example table types:
+
+1. Prefix: longest matching prefix of header values
+2. Exact: exact match
+
+Implementation: [fwdport](forwarding/fwdtable)
+
+### saiserver
+
+The saiserver is a gRPC API generated from the [SAI api](https://github.com/opencomputeproject/SAI). It provides a simpler API for programming the dataplane.
+The saiserver package implements the API by calling the corresponding forwarding API. This package also configures the forwarding pipeline, by setting up the
+required tables.
+
+A simple example: the CreateRouteEntry RPC adds a entry to the FIB. This looks something like:
+
+1. Compute forwarding actions
+     1. Update the next hop ip metadata field.
+     2. If the next hop is port: transmit action.
+     3. If next hop is next hop: update packet metadata next hop id, lookup next-hop table.
+     4. If next hop is a group: update packet metadata next hop group, lookup next-hop-group table.
+2. Figure out the address family of the prefix.
+3. Add forwarding table entry to correct fib: prefix entry with fields vrf and next hop ip and above action.
+
+### dplanerc
+
+The dplanerc implements a gNMI reconciler based on [reconciler package](../gnmi/reconciler/reconciler.go). These reconcilers convert from OpenConfig to
+saiserver proto messages to support configuring the dataplane from gNMI.
+
+Note: The fib uses a custom, simple proto message because OpenConfig does not model configured forwarding tables (only state).
+
+### standalone
+
+The dataplane standalone is meant to be used by a NOS to configure a dataplane using the C SAI API. (eg SONiC).
+The current design runs the dataplane separately from the NOS and all communication is over gRPC.
+
+* The sai directory contains generated C++ code that converts the C SAI API to protobuf API and performs the RPCs.
+  * Note: the common.cc and common.h contain hand-crafted conversion funcs.
+* The packetio package receives packets punted to the CPU port and transmits to the correct interface. It programs the dataplane with the IP assigned to the local interfaces.
+* The lucius directory is the Go main that starts the dataplane RPC server.
+* The entrypoint.cc is compiled to a shared library that can be loaded in the NOS.
+
+### apigen
+
+These packages generate the protobuf and C++ source and headers based on the SAI headers.
+
+## How to add a new feature
```

Comment:
this is pretty confusing, but I dont think I'm the intended audience here

it might help if more section headers were added:

add dataplane support

etc.

Reply:
rewrote this whole section a bit for additional clarity